### PR TITLE
Migrate aggregation of validation status, execution time, and task executions to Athena

### PR DESCRIPTION
**Pull request by kathy-t**
**Description**
This PR migrates the aggregation of validation status, execution time, and task executions to Athena.

For validation status, this involved writing a new query.

For execution time, this involved using the new `executiontimeseconds` property introduced by https://github.com/dockstore/dockstore/pull/5842.

For task executions, this involved writing a query that creates a single row representing a workflow execution from an array of tasks. It follows the same rules as https://github.com/dockstore/dockstore-support/blob/bf345fd1aeb613e913e98165fe8c4c969261c05b/metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/RunExecutionAggregator.java#L24-L30
- I've noted in the relevant file that the aggregation of task execution status is slightly different because it doesn't translate well to SQL. What we did before was set the status as successful if all task execution statuses were successful, otherwise set it as the most frequently occurring failed status. In Athena, I set it as successful if all task execution statuses were successful, otherwise set it as `FAILED`. We currently don't ingest any task execution statuses, but we may want to re-visit how we aggregate task executions when we do.

Also did some slight clean up work to handle exceptions better and moved some functions around, so I recommend hiding whitespace when reviewing.

I aggregated all executions in staging using Java and Athena (skipping posting to Dockstore):
- Java: PT58M0.370870455S
- Athena: PT59M46.777237722S

I suspect that the bottleneck for Athena is that we execute the query for validation status even if there are no validation executions. I will ponder this for a bit to see if there's a way to avoid that.

Athena is still faster when it comes to aggregating purely run executions.

**Review Instructions**
not user facing

**Issue**
[SEAB-6412](https://ucsc-cgl.atlassian.net/browse/SEAB-6413) - migrate validation status and execution time
[SEAB-6414](https://ucsc-cgl.atlassian.net/browse/SEAB-6414) - migrate task executions

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install` in the project that you have modified (until https://ucsc-cgl.atlassian.net/browse/SEAB-5300 adds multi-module support properly)
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] If you are changing dependencies, check with dependabot to ensure you are not introducing new high/critical vulnerabilities
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead. 

[SEAB-6412]: https://ucsc-cgl.atlassian.net/browse/SEAB-6412?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/RunExecutionAthenaAggregator.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                                .and(VERSION_FIELD.eq(inline(partition.version())))));
+        // This query created a workflow execution from each array of tasks. This will be unioned with the actual workflow executions submitted
+        final CommonTableExpression<Record7<String, String, String, Integer, Double, Integer, String>> runExecutionsFromTasksTable = name("runexecutionfromtasks")
+                .fields(PLATFORM_FIELD.getName(), DATE_EXECUTED_FIELD.getName(), EXECUTION_STATUS_FIELD.getName(), EXECUTION_TIME_SECONDS_FIELD.getName(),
+                        MEMORY_REQUIREMENTS_GB_FIELD.getName(), CPU_REQUIREMENTS_FIELD.getName(), COST_FIELD.getName())
+                .as(select(PLATFORM_FIELD,
+                        field("array_min(transform(taskexecutions, t -> t.dateexecuted))", String.class).as(DATE_EXECUTED_FIELD),
+                        // If all tasks are successful, set the workflow execution status as successful. Otherwise, assume failed
+                        field("case when all_match(transform(taskexecutions, t -> t.executionstatus), t -> t = 'SUCCESSFUL') then 'SUCCESSFUL' else 'FAILED' end", String.class).as(EXECUTION_STATUS_FIELD),
```
Comment (kathy-t) [inline metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/RunExecutionAthenaAggregator.java:115]: This is where the task execution status aggregation differs from the Java version:
https://github.com/dockstore/dockstore-support/blob/bf345fd1aeb613e913e98165fe8c4c969261c05b/metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/ExecutionStatusAggregator.java#L49-L53

There's no clean way to replicate it in SQL so if all task statuses are not successful, I set it to a generic fail.

**Comment by denis-yuen** [comment]
> I suspect that the bottleneck for Athena is that we execute the query for validation status even if there are no validation executions. I will ponder this for a bit to see if there's a way to avoid that.

:+1:

File: metricsaggregator/src/main/java/io/dockstore/metricsaggregator/MetricsAggregatorAthenaClient.java [hunk 2]
```diff
@@ -75,144 +77,59 @@ public void aggregateMetrics(List<S3DirectoryInfo> s3DirectoriesToAggregate, Ext
...
-
-            for (GetQueryResultsResponse result : getQueryResultsIterable.get()) {
-                // Create a map of column name to column value index. Only needs to be created once because the other pages should have the same columns
-                if (columnNameToColumnIndex.isEmpty()) {
-                    List<ColumnInfo> columnInfoList = result.resultSet().resultSetMetadata().columnInfo();
-                    for (int i = 0; i < columnInfoList.size(); ++i) {
-                        ColumnInfo columnInfo = columnInfoList.get(i);
-                        columnNameToColumnIndex.put(columnInfo.name(), i);
-                    }
+        //LOG.info("Running SQL query:\n{}", query);
```
Comment (denis-yuen) [inline metricsaggregator/src/main/java/io/dockstore/metricsaggregator/MetricsAggregatorAthenaClient.java:110]: Could erase or just set to `debug`

File: metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/AthenaAggregator.java [hunk 3]
```diff
@@ -1,182 +1,195 @@
...
+                )
+                """, tableName, metricsBucketName, entityProjectionValues, registryProjectionValues, platformProjectionValues);
 
-    public Optional<Integer> getCountColumnValue(QueryResultRow queryResultRow) {
-        Optional<Integer> countColumnValue = queryResultRow.getColumnValue(getCountColumnName()).map(Integer::valueOf);
-        if (countColumnValue.isPresent() && countColumnValue.get() == 0) { // There were 0 non-null column values
-            return Optional.empty();
+        try {
+            metricsAggregatorAthenaClient.executeQuery(query);
+        } catch (Exception e) {
```
Comment (denis-yuen) [inline metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/AthenaAggregator.java:191]: Really general, without looking at the code are there different kinds of exceptions that could be addressed differently?

File: metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/RunExecutionAthenaAggregator.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        final CommonTableExpression<Record7<String, String, String, Integer, Double, Integer, String>> executionsTable = name("executions")
+                .fields(PLATFORM_FIELD.getName(), DATE_EXECUTED_FIELD.getName(), EXECUTION_STATUS_FIELD.getName(), EXECUTION_TIME_SECONDS_FIELD.getName(),
+                        MEMORY_REQUIREMENTS_GB_FIELD.getName(), CPU_REQUIREMENTS_FIELD.getName(), COST_FIELD.getName())
+                .as(select(PLATFORM_FIELD,
+                        field("unnested." + DATE_EXECUTED_FIELD, String.class),
+                        field("unnested." + EXECUTION_STATUS_FIELD, String.class),
+                        field("unnested." + EXECUTION_TIME_SECONDS_FIELD, Integer.class),
+                        field("unnested." + MEMORY_REQUIREMENTS_GB_FIELD, Double.class),
+                        field("unnested." + CPU_REQUIREMENTS_FIELD, Integer.class),
+                        field("unnested." + COST_FIELD, String.class))
```
Comment (denis-yuen) [inline metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/RunExecutionAthenaAggregator.java:101]: Thought for the future, wonder if we'll have to be more sophisticated about cost. A string also may make difficult to process via Athena but we may have to deal with different currencies.
Can deal with it when we have (the good problem) to deal with it

**Review by denis-yuen — COMMENTED** [review]
Neat!
Curious what the other comments are, but looking good

File: metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/RunExecutionAthenaAggregator.java (hunk 4 above)
Comment (kathy-t) [reply metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/RunExecutionAthenaAggregator.java:101]: Specifying this as a `String` doesn't affect the resulting query for cost. This query is just for flattening the RunExecutions array and this line will resolve to `unnested.cost` and will have the form of a `Cost` object (with value and currency).

The cost aggregator currently only takes into account the cost value, but like you mentioned, we may get more sophisticated with our cost aggregation later. When this happens, the change will occur somewhere in this file so it takes into account `cost.currency` for example  https://github.com/dockstore/dockstore-support/blob/bf345fd1aeb613e913e98165fe8c4c969261c05b/metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/CostAthenaAggregator.java#L18

File: metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/AthenaAggregator.java (hunk 3 above)
Comment (kathy-t) [reply metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/AthenaAggregator.java:191]: I wanted to catch Athena related errors that can occur during query execution. Looking at the javadoc for the athena client methods called, catching AwsServiceException and SdkClientException should cover it, so I'll change it to that so unrelated errors can be propogated

File: metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/ExecutionTimeAthenaAggregator.java [hunk 5]
```diff
@@ -1,22 +1,38 @@
...
+    public ExecutionTimeAthenaAggregator(MetricsAggregatorAthenaClient metricsAggregatorAthenaClient, String tableName) {
+        super(metricsAggregatorAthenaClient, tableName);
+        this.addSelectFields(getStatisticSelectFields());
+    }
 
     @Override
-    public String getMetricColumnName() {
-        return "executiontime";
+    String getMetricColumnName() {
+        return "executiontimeseconds";
```
Comment (coverbeck) [inline metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/ExecutionTimeAthenaAggregator.java:20]: Nit, might be nice to have a "constants" file with this and other column names, so you can see in one place what all the column names are.

Update, I guess these are already [these values](https://github.com/dockstore/dockstore-support/blob/078a17204bd50922fe65d65ae3dde3db620f6a79/metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/RunExecutionAthenaAggregator.java#L36)?

**Review by coverbeck — APPROVED** [review]

File: metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/ExecutionTimeAthenaAggregator.java (hunk 5 above)
Comment (kathy-t) [reply metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/ExecutionTimeAthenaAggregator.java:20]: > Update, I guess these are already [these values](https://github.com/dockstore/dockstore-support/blob/078a17204bd50922fe65d65ae3dde3db620f6a79/metricsaggregator/src/main/java/io/dockstore/metricsaggregator/helper/RunExecutionAthenaAggregator.java#L36)?

Oh yes they are, I'll update these usages to the constants

**Review by svonworl — APPROVED** [review]

**Pull request merged**